File: src/apps/api/app/pipeline/text_animation_phases.py
"""Absolute-time motion used by authored text preview/export.

Swift mirror: KriaMediaEngine/TextAnimationPhases.swift. Animation does not
retime the containing item. Short windows share their available time equally.
"""

from dataclasses import dataclass
from math import floor, pi, sin

from app.agents._schemas.text_animation_phases import TextAnimationPhases


@dataclass(frozen=True)
class TextPhaseSample:
    alpha: float = 1
    scale: float = 1
    x: float = 0
    y: float = 0
    reveal: float = 1


def sample_text_phases(
    phases: TextAnimationPhases, time: float, duration: float
) -> TextPhaseSample:
    if duration <= 0 or time < 0 or time >= duration:
        return TextPhaseSample(alpha=0, reveal=0)
    phase_duration = min(0.4 / phases.speed, duration / 2)
    alpha, scale, x, y, reveal = 1.0, 1.0, 0.0, 0.0, 1.0
    for effect, progress in (
        (phases.entrance, min(1.0, time / phase_duration)),
        (phases.exit, min(1.0, (duration - time) / phase_duration)),
    ):
        eased = 1 - (1 - progress) ** 3
        if effect == "fade":
            alpha *= eased
        elif effect == "pop":
            scale *= 0.7 + 0.3 * eased
            alpha *= eased
        elif effect == "slide":
            x += 40 * (1 - eased)
            alpha *= eased
        elif effect == "typewriter":
            reveal = min(reveal, progress)
    wave = sin(2 * pi * time * phases.speed / 1.2)
    if phases.loop == "pulse":
        scale *= 1 + 0.04 * wave
    elif phases.loop == "bounce":
        y -= 12 * abs(wave)
    elif phases.loop == "float":
        y -= 8 * wave
    return TextPhaseSample(alpha, scale, x, y, reveal)


def visible_grapheme_count(count: int, reveal: float) -> int:
    return min(count, max(0, floor(count * reveal + 1e-9)))

File: src/apps/ios/Packages/KriaMediaEngine/Sources/KriaMediaEngine/TextAnimationPhases.swift
import Foundation

/// Deterministic output-clock sampling shared by native preview and export.
public extension TextAnimationPhases {
    func sample(time: Double, duration: Double) throws -> TextTransformSample {
        try validate()
        guard time.isFinite, duration.isFinite, duration > 0 else { throw RecipeError.invalidTimeline }
        guard time >= 0, time < duration else {
            return TextTransformSample(alpha: 0, scale: 1, xTranslate: 0, yTranslate: 0, revealProgress: 0)
        }
        let edgeDuration = min(0.4 / speed, duration / 2)
        var alpha = 1.0, scale = 1.0, x = 0.0, y = 0.0, reveal = 1.0
        for (effect, progress) in [(entrance, min(1, time / edgeDuration)), (exit, min(1, (duration - time) / edgeDuration))] {
            let eased = 1 - pow(1 - progress, 3)
            switch effect {
            case .none: break
            case .fade: alpha *= eased
            case .pop: scale *= 0.7 + 0.3 * eased; alpha *= eased
            case .slide: x += 40 * (1 - eased); alpha *= eased
            case .typewriter: reveal = min(reveal, progress)
            }
        }
        let wave = sin(2 * Double.pi * time * speed / 1.2)
        switch loop {
        case .none: break
        case .pulse: scale *= 1 + 0.04 * wave
        case .bounce: y -= 12 * abs(wave)
        case .float: y -= 8 * wave
        }
        return TextTransformSample(alpha: alpha, scale: scale, xTranslate: x, yTranslate: y, revealProgress: reveal)
    }
    static func visiblePrefix(_ text: String, progress: Double) -> String {
        guard progress.isFinite else { return "" }
        return String(text.prefix(min(text.count, max(0, Int(floor(Double(text.count) * progress + 1e-9))))))
    }
}
